mono/metadata/object.c

```diff
@@ -3966,25 +3967,25 @@ mono_runtime_get_main_args (void)
  * \returns a \c MonoArray with the arguments passed to the main
  * program. On failure returns NULL and sets \p error.
  */
-MonoArray*
-mono_runtime_get_main_args_checked (MonoError *error)
+MonoArrayHandle*
+mono_runtime_get_main_args_checked_handle (MonoError *error)
 {
-	MonoArray *res;
+	HANDLE_FUNCTION_ENTER ();
+	MonoArrayHandle *res;
 	int i;
 	MonoDomain *domain = mono_domain_get ();
-
 	error_init (error);
 
-	res = (MonoArray*)mono_array_new_checked (domain, mono_defaults.string_class, num_main_args, error);
-	return_val_if_nok (error, NULL);
+	res = (MonoArrayHandle*) mono_array_new_handle(domain, mono_defaults.string_class, num_main_args, error);
+	mono_error_assert_ok (error);
 
 	for (i = 0; i < num_main_args; ++i) {
-		MonoString *arg = mono_string_new_checked (domain, main_args [i], error);
-		return_val_if_nok (error, NULL);
-		mono_array_setref (res, i, arg);
+		MonoStringHandle arg_string = mono_string_new_handle (domain, main_args [i], error);
```

Comment:
This loop allocates handles in proportion to user input.  We typically try to keep it to a fixed number of handles per function.

You will most likely need to lift the loop body into a static helper function that uses `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN_*` and allocates a single handle for the "current" arg.

Take a look at, for example, `add_event_other_methods_to_array` in icall.c for an example of such a loop body.